build.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use regex::Regex;
+use std::env;
+use std::fs::read_dir;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::process::Command;
+
+fn main() {
+    // This build script creates files in the `src` directory. Since that is
+    // outside Cargo's OUT_DIR it will cause an error when this crate is used
+    // as a dependency. Therefore, the user must opt-in to regenerating the
+    // Rust files.
+    if env::var_os("CARGO_FEATURE_REGENERATE").is_none() {
+        println!("cargo:rerun-if-changed=build.rs");
+        return;
+    }
+
+    let buf_lib = BufferLib::from_env_vars();
+    if buf_lib == BufferLib::Prost {
+        unimplemented!("Prost support is not yet implemented");
+    }
+
+    check_protoc_version();
+
+    let file_names: Vec<_> = read_dir("proto")
+        .expect("Couldn't read proto directory")
+        .map(|e| {
+            format!(
+                "proto/{}",
+                e.expect("Couldn't list file").file_name().to_string_lossy()
+            )
+        })
+        .collect();
+    let file_names: Vec<_> = file_names.iter().map(|s| &**s).collect();
+
+    for f in &file_names {
+        println!("cargo:rerun-if-changed={}", f);
+    }
+
+    if buf_lib == BufferLib::Protobuf {
+        generate_protobuf_files(file_names);
+    }
+
+    if buf_lib == BufferLib::Protobuf {
```

Comment:
Why aren't these `if`s merged together?

Reply:
good catch, I just missed the opportunity when I moved `mod_names`